# __Ensemble Learning__

## __Agenda__

In this lesson, let's cover the following concepts with examples:
- Introduction to ensemble learning
    * Goals of ensemble learning
    * Importance of ensemble learning
- Categories in ensemble learning
    * Sequential ensemble technique
    * Parallel ensemble technique
- Simple techniques for ensemble learning
    * Max voting
    * Averaging
    * Weighted averaging
- Advanced techniques for ensemble learning
    * Bagging (bootstrap aggregating)
        * Advantages of bagging
        * Disadvantages of bagging
        * Out-of-bag (OOB) concept
    * Boosting
        * Advantages of boosting
        * Disadvantages of boosting
    * Stacking
        * Advantages of stacking
        * Disadvantages of stacking
    * Blending ensemble
        * Advantages of blending ensemble
        * Disadvantages of blending ensemble
- AdaBoost boosting algorithms

## __1. Introduction to Ensemble Learning__

Ensemble learning combines multiple models to enhance the overall performance of machine learning algorithms. The fundamental principle of ensemble learning is combining predictions from multiple individual models to produce a more accurate and robust prediction than any single model.

### __1.1 Goals of Ensemble Learning__

- Enhance predictive accuracy by combining multiple models.
- Improve model robustness and generalization performance.


![link text](https://labcontent.simplicdn.net/data-content/content-assets/Data_and_AI/ML/Lesson_07/ensemble_learning.png)

__Example:__ Imagine you are deciding whether to go outside based on weather predictions. Instead of relying on just one weather forecast, you check multiple sources, like the weather app, a meteorologist's report, and a weather website. You then combine these predictions to make your decision.

This process mirrors ensemble learning, where combining multiple models (or weather forecasts) results in a more accurate prediction (or decision) than any single model could offer.

 ### __1.2 Importance of Ensemble Learning__

- Improves prediction accuracy by combining diverse models
- Enhances model resilience and robustness against uncertainties
- Mitigates biases and errors present in individual models
- Captures a wide range of perspectives to achieve better performance
- Provides reliable and robust forecasts across various domains, ensuring more dependable outcomes

## __2. Categories in Ensemble Learning__

Ensemble learning can be broadly classified into two categories:
- Sequential ensemble technique
- Parallel ensemble technique

### __2.1 Sequential Ensemble Technique__

These techniques train models sequentially, with each model attempting to correct its predecessor's errors. This technique focuses on improving the overall performance of the ensemble by iteratively refining predictions. An example of this approach is boosting.

__Note:__ It typically employs weak learners as base estimators because these learners initially have higher error rates.

The steps involved in the sequential ensemble technique depend on the data structure and the requirements of the application. They are:
1. __Base Model Selection__: Choose the initial model.
2. __Sequential Training__: Train models one after the other.
3. __Error Correction__: Each model learns from previous mistakes.
4. __Prediction Refinement__: Iteratively refine predictions.
5. __Combining Predictions__: Combine predictions from all models.
6. __Evaluation__: Assess ensemble performance using metrics.


![link text](https://labcontent.simplicdn.net/data-content/content-assets/Data_and_AI/ML/Lesson_07/Sequential_Ensemble_Technique.png)

__Note:__ If all four base models are of the same type, it is considered homogeneous. If they are different, it is considered heterogeneous.

The diagram above shows the training data divided into four samples, with each sample trained on a distinct base model. Insights gained from model M1 are passed to model M2 alongside sample S2. M2 then adjusts its weights and biases based on the outcomes of M1. This process repeats for models 2, 3, and 4. Finally, all the learners are combined using a weighted averaging strategy.

The summation sign ($ ∑ $) indicates the function that adjusts the model to improve its overall performance.

__The sequential ensemble technique is employed when dealing with:__
- Complex relationships between input features and the target variable
- Diverse data types, including numerical, categorical, and textual data
- High-dimensional datasets with numerous features
- Imbalanced datasets, where skewed class distributions pose classification challenges
- Incremental updates are needed to adapt models gradually to evolving data over time

### __2.2 Parallel Ensemble Technique__

The parallel ensemble technique concurrently trains models. They combine the predictions from multiple models to improve the final output. Bagging and Random Forest algorithms are examples of parallel ensemble techniques.

__Note:__  It employs stronger learners as base estimators.

The steps in the parallel ensemble technique are:

1. __Data Partitioning__: Divide the dataset into subsets
2. __Model Training__: Train models concurrently on subsets
3. __Prediction__: Models make independent predictions
4. __Combining Predictions__: Aggregate predictions using techniques like voting or averaging
5. __Evaluation__: Assess ensemble performance using metrics


![link text](https://labcontent.simplicdn.net/data-content/content-assets/Data_and_AI/ML/Lesson_07/Parallel_Ensemble_Technique.png)

In the above diagram, the training data is divided into four samples, labeled S1 to S4, each trained on a distinct base model (M1 to M4). Unlike the sequential ensemble technique, the data in the base learners is independent. This independence of base learners significantly reduces the error due to the application of averages.

The summation sign ($ ∑ $) indicates the aggregated model with improved performance, which has been learned from all independent base models, M1 to M4.

The parallel ensemble technique is used for:

- Making training models faster in big datasets by doing many things at once
- Training many models at once by having lots of computer power
- Processing different kinds of data and features all together
- Efficiently using many computers together for big tasks
- Making better predictions by getting different behaviors from models trained separately

## __3. Simple Techniques for Ensemble Learning__

These techniques combine predictions from multiple base models without using any complex algorithms. They aim to improve accuracy and robustness by leveraging diverse model perspectives.
They are typically applied to straightforward tasks like classification, regression, and basic pattern recognition.

### __3.1 Max Voting__

Max Voting is an ensemble learning classification technique. It involves multiple models, making predictions for each data point. Each model's prediction is considered a __vote__. The final prediction is determined by the majority vote among the models.

__Example:__
    
- Max Voting ensemble works on breast cancer classification by combining predictions from multiple individual classifiers, such as Logistic Regression, Decision Tree, and Support Vector Machine.
- Each classifier provides its prediction for whether a given sample belongs to a certain class. The voting ensemble then aggregates these predictions using a voting mechanism.
- The final prediction is determined based on the most commonly predicted class among all classifiers.
- This approach leverages the collective wisdom of diverse models to improve overall prediction accuracy and robustness in breast cancer classification tasks.    

Let's look at an implementation of max voting using a breast cancer classification dataset.

In [ ]:
# Import necessary libraries
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import r2_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC

In [ ]:
# Load the dataset from the specified URL and assign column names to the DataFrame
data = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/breast-cancer-wisconsin.data', header=None)
data.columns = ['Sample code', 'Clump Thickness', 'Uniformity of Cell Size', 'Uniformity of Cell Shape',
                'Marginal Adhesion', 'Single Epithelial Cell Size', 'Bare Nuclei', 'Bland Chromatin',
                'Normal Nucleoli', 'Mitoses', 'Class']

In [ ]:
# Drop the 'Sample code' column as it's not required for prediction
data.drop(['Sample code'], axis=1, inplace=True)

# Replace '?' with 0 to handle missing data, ensuring numerical analysis accuracy and preventing calculation errors.
data = data.replace('?', np.nan)

# Convert the 'Bare Nuclei' column to an integer type to ensure a uniform data type across the column,
# which is essential for consistent data manipulation and analysis.
data['Bare Nuclei'] = pd.to_numeric(data['Bare Nuclei']).fillna(0).astype('int64')

In [ ]:
# Display information about the dataset
data.info()

# Extract features (X) and target variable (y) from the dataset
values = data.values
X = values[:, 0:9]
y = values[:, 9]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 699 entries, 0 to 698
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype
---  ------                       --------------  -----
 0   Clump Thickness              699 non-null    int64
 1   Uniformity of Cell Size      699 non-null    int64
 2   Uniformity of Cell Shape     699 non-null    int64
 3   Marginal Adhesion            699 non-null    int64
 4   Single Epithelial Cell Size  699 non-null    int64
 5   Bare Nuclei                  699 non-null    int64
 6   Bland Chromatin              699 non-null    int64
 7   Normal Nucleoli              699 non-null    int64
 8   Mitoses                      699 non-null    int64
 9   Class                        699 non-null    int64
dtypes: int64(10)
memory usage: 54.7 KB


In [ ]:
# Import required libraries
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import MinMaxScaler

# Create an instance of SimpleImputer to fill in missing values
imputer = SimpleImputer()

# Impute the missing values in the dataset and obtaining imputed data
imputeData = imputer.fit_transform(values)

# Create an instance of MinMaxScaler to scale features to a range of (0, 1)
scaler = MinMaxScaler(feature_range=(0, 1))

# Scale the imputed data features to a range of (0, 1) to obtain normalized data
normalizedData = scaler.fit_transform(imputeData)

In [ ]:
# Import libraries
from sklearn.ensemble import VotingClassifier
from sklearn import model_selection

# Set up KFold cross-validation with 10 splits
kfold = model_selection.KFold(n_splits=10, random_state=7, shuffle=True)

# Initialize list for base models
estimators = []

# Add Logistic Regression model to estimators
model1 = LogisticRegression()
estimators.append(('logistic', model1))

# Add Decision Tree model to estimators
model2 = DecisionTreeClassifier()
estimators.append(('cart', model2))

# Add Support Vector Machine model to estimators
model3 = SVC()
estimators.append(('svm', model3))

# Create Voting Classifier ensemble with estimators
ensemble = VotingClassifier(estimators)

# Evaluate ensemble model using cross-validation
results = model_selection.cross_val_score(ensemble, X, y, cv=kfold)

# Print mean accuracy score of ensemble model
print(results.mean())

0.9642028985507245


The max voting ensemble's 96.4% accuracy rate demonstrates its remarkable effectiveness in predicting breast cancer classifications. By combining predictions from multiple individual classifiers, this ensemble leverages the collective insights of diverse models to achieve highly accurate results.

### __3.2 Averaging__

The averaging technique is used mainly for regression problems. In this method, the predictions of multiple models are averaged to obtain the final prediction. This technique helps reduce variance and produces a more stable prediction.

__Example:__

- Averaging ensemble works by combining the predictions from multiple classifiers trained on the dataset.
- These classifiers, such as Logistic Regression, Decision Tree, and Support Vector Machine, provide individual predictions for each data point.
- Then it takes the average of these predictions to form a final prediction.
- This approach helps to mitigate biases and uncertainties inherent in individual models, leading to a more accurate prediction of breast cancer classifications.

In [ ]:
# Initialize classifiers: Logistic Regression, Decision Tree, and Support Vector Classifier (SVC)
LogRefg_clf = LogisticRegression()
Dtree_clf = DecisionTreeClassifier()
svc_slf = SVC()

# Train classifiers on the normalized dataset
LogRefg_clf.fit(X, y)
Dtree_clf.fit(X, y)
svc_slf.fit(X, y)

# Generate predictions for each classifier
LogRefg_pred = LogRefg_clf.predict(X)
Dtree_pred = Dtree_clf.predict(X)
svc_pred = svc_slf.predict(X)

# Average the predictions from all classifiers
avg_preds = np.mean([LogRefg_pred, Dtree_pred, svc_pred], axis=0)

# Evaluate the accuracy of the averaged predictions using R-squared score
acc = r2_score(y, avg_preds)

# Print the R-squared score
print(acc)

0.9521523008812143


The averaging ensemble achieves an accuracy of 94%, slightly lower than the max voting ensemble. However, this accuracy rate still underscores the effectiveness of the averaging ensemble in breast cancer classification.

### __3.3 Weighted Averaging__

Weighted averaging assigns different weights to each model’s prediction, reflecting its importance or reliability. The final prediction is a weighted average, which can be more effective than simple averaging because it considers each model's performance.

__Example:__

- In medical diagnosis, weighted ensemble learning combines results from various diagnostic tests with different reliabilities.
- Each test's result is weighted based on its importance. For instance, blood tests, imaging scans, and physical examinations' results are combined to determine a patient's diagnosis accurately and reliably.

By weighting and averaging results from multiple diagnostic tests, you can determine the final diagnosis, thereby enhancing diagnostic accuracy and reliability. This highlights the significance of weighted averaging for improving decision-making.

## __4. Advanced Techniques for Ensemble Learning__

These techniques use complex methods to merge predictions from many individual models. It aims for better predictive accuracy. They are employed for complex problems where high predictive accuracy is crucial, such as in financial forecasting, medical diagnosis, and natural language processing applications.

### __4.1 Bagging__

Bagging or Bootstrap Aggregating enhances the stability and accuracy of machine learning algorithms. This method creates multiple subsets from the original dataset, known as bootstrap samples, by selecting data points with replacement. Each subset trains a separate model.

For final predictions, Bagging combines the individual models' outputs by using voting (for classification tasks) or averaging (for regression tasks). This approach effectively reduces variance and helps prevent overfitting.

__Note:__ Random Forest is a popular algorithm that utilizes bagging by training multiple decision trees on different bootstrap samples and combining their predictions.


![link text](https://labcontent.simplicdn.net/data-content/content-assets/Data_and_AI/ML/Lesson_07/bagging.png)

__Note:__ The numbering shown in the above image outlines the specific workflow for bagging.

In [ ]:
# Import required libraries
import pandas as pd
from sklearn import model_selection
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [ ]:
# Load the Pima Indians Diabetes dataset using the given URL and assign the names

url="https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv"
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
data = pandas.read_csv(url, names=names)

# Check the head of the data using the head() method
data.head()

__Observations:__
- This is the head of the dataset.

In [ ]:
# Create a feature matrix X and a target vector Y

# Assign all the rows up to the 8th column of the data to X
array = data.values
X = array[:,0:8]

# Assign the 9th column of the same data to Y
Y = array[:,8]

In [ ]:
# Create a Decision Tree Classifier

# Set up the k-fold cross-validation with n_splits to 10
kfold = model_selection.KFold(n_splits=10)

# Create a decision tree classifier named cart
cart = DecisionTreeClassifier()

# The estimator will be set to 100, which is equal to the number of trees
num_trees = 100

In [ ]:
# Apply the bagging technique
# Create a bagging classifier with the decision tree classifier (cart) as the base estimator
# Ensure oob_score is enabled
model = BaggingClassifier(estimator=cart,n_estimators=num_trees, random_state=12, oob_score=True)

# Perform cross-validation using the model, feature matrix (X), target vector (Y), and specified number of folds (cv)
results = model_selection.cross_val_score(model, X, Y, cv=kfold)

# Print the cross-validation results
print(results)

__Observations:__
- The results array contains the cross-validation scores for each fold of the data.
- The cross-validation scores indicate the performance of the BaggingClassifier model on different subsets of the data.
- These scores can assess the generalization ability and performance stability of the BaggingClassifier model across multiple iterations of the training data.

In [ ]:
# Calculate the Mean Cross-Validation Score using .mean() function
mean_score = np.mean(results)

print("Mean Cross-Validation Score:", mean_score)

- The provided code illustrates a bagging implementation with a DecisionTreeClassifier base estimator. It employs KFold cross-validation with 10 folds and applies BaggingClassifier with 100 decision tree estimators to each fold.
- The resultant mean cross-validation score of 0.76688 demonstrates the ensemble's effectiveness in enhancing predictive performance compared to a single decision tree classifier.

### __4.1.1 Advantages of Bagging__

- It reduces the risk of overfitting by training on various subsets of the data.
- It improves the model's accuracy and stability.
- It works well with complex models that tend to overfit.

### __4.1.2 Disadvantages of Bagging__

- It increases computational complexity because it requires training multiple models.
- It may not significantly improve performance if the base models are already biased.

### __4.1.3  Out-Of-Bag (OOB) Concept__

In bagging, __out-of-bag (OOB)__ data refers to instances not included in the bootstrap sample for training a specific base model. Since the process trains each model on a random subset of the original data, it naturally leaves out some instances.

These OOB instances serve as a built-in validation set for each model, allowing for an assessment of the model's performance on unseen data points. You can compute the OOB error, which offers an efficient way to evaluate the model's generalization ability.

This method eliminates the need for additional validation data or cross-validation procedures.


![link text](https://labcontent.simplicdn.net/data-content/content-assets/Data_and_AI/ML/Lesson_07/Out_Of_Bag_Concept.png)

In the above image, Sample 1 lacks Rat and Cow, whereas Sample 3 includes all animals from the primary training set. Data points were randomly chosen with replacement during sample creation, and those omitted from a particular sample are identified as out-of-bag (OOB) points.

In [ ]:
# Perform cross-validation and compute OOB error
oob_error = 1 - model.fit(X, Y).oob_score_

print("Out-of-Bag Error:", oob_error)

__Observations:__
- The code computes the out-of-bag (OOB) error for a BaggingClassifier model using cross-validation.
- It calculates the OOB error by subtracting the OOB score from 1.
- The **OOB score** represents the model's accuracy on out-of-bag samples.
- The OOB error provides insight into the model's performance on unseen data points, indicating the misclassification rate of the model.

### __4.2 Boosting__

Boosting is a sequential ensemble technique where each model corrects its predecessor's mistakes. The main idea is to train models sequentially, each trying to correct the errors of the previous model.

The final prediction is a weighted sum of all the models' predictions. Boosting significantly reduces bias and variance, leading to more accurate models.

__Note:__ Popular boosting algorithms include AdaBoost, Gradient Boosting Machines (GBM), and XGBoost.


![link text](https://labcontent.simplicdn.net/data-content/content-assets/Data_and_AI/ML/Lesson_07/boosting.png)

__Note:__ The numbering shown in the above image specifically outlines the workflow for boosting.

In [ ]:
# Load the breast cancer dataset and create a DataFrame df
df = pd.DataFrame(load_breast_cancer()['data'], columns=load_breast_cancer()['feature_names'])

# Assign the feature names of the dataset to columns and assign the target column to y
df['y'] = load_breast_cancer()['target']

# Using the head() method, you can check the first 5 rows of the dataset
df.head(5)

NameError: name 'pd' is not defined

In [ ]:
# Check the data types and find information about the dataset
df.info()

__Observation:__
- There are no missing or null values.

In [ ]:
# Perform K-Fold Cross-Validation and Fit the Model
# Define X and y
# Assign the DataFrame df, except the target column, to X
X = df.drop('y', axis=1)

# Assign the y column to variable y
y = df['y']

In [ ]:
# Initialize lists to store evaluation metrics
accuracy_collection = []
recall_collection = []
precision_collection = []

# Create a K-fold object with 5 splits
kf = KFold(n_splits=5, random_state=42, shuffle=True)

# Iterate over the K-fold splits
for train_index, val_index in kf.split(X):
    # Split the data into training and validation sets
    X_train, X_val = X.iloc[train_index], X.iloc[val_index]
    y_train, y_val = y.iloc[train_index], y.iloc[val_index]

    # Initialize a GradientBoostingClassifier object
    model = GradientBoostingClassifier()

    # Fit the model on the training data
    model.fit(X_train, y_train)

    # Make predictions on the validation set
    predictions = model.predict(X_val)

    # Calculate accuracy and store in accuracy_collection
    accuracy = accuracy_score(y_val, predictions)
    accuracy_collection.append(accuracy)

# Print accuracy for each fold
for i, acc in enumerate(accuracy_collection, start=1):
  print(f'Accuracy of fold {i}: {acc}')

# Print average accuracy across all folds
print(f'Average Accuracy: {np.mean(accuracy_collection)}')

In [ ]:
# Generate a classification report for the validation set predictions
report = classification_report(y_val, predictions, output_dict=True)

# Extract recall and precision values for class label '1'
recall = report['1']['recall']
precision = report['1']['precision']

# Append recall and precision values to their respective collections
recall_collection.append(recall)
precision_collection.append(precision)

# Calculate the overall recall and precision by taking the mean of the collections
overall_recall = np.mean(recall_collection)
overall_precision = np.mean(precision_collection)

# Print the overall recall and precision values
print(f'Overall Recall: {overall_recall}')
print(f'Overall Precision: {overall_precision}')

__Observations:__

- The accuracy is 0.96.
- Precision and recall do not have much difference between them.

### __4.2.1 Advantages of Boosting__

- It enhances accuracy by reducing both bias and variance significantly.
- It is adaptable and compatible with various types of models.
- It is efficient in complex scenarios where simple models struggle.


### __4.2.2 Disadvantages of Boosting__

- It is more susceptible to overfitting compared to Bagging when dealing with noisy data.
- It demands more computational resources as models are trained sequentially.
- It requires careful parameter tuning to prevent overfitting.

### __4.3 Stacking__

Stacking combines multiple classification or regression models via a meta-classifier or a meta-regressor. After training the base models on the complete dataset, the meta-model trains on the base models' outputs as features.

This technique leverages the strengths of each base model and can achieve higher accuracy.


![link text](https://labcontent.simplicdn.net/data-content/content-assets/Data_and_AI/ML/Lesson_07/stacking.png)

__Note:__ The numbering shown in the above image specifically outlines the workflow for stacking.

- Stacking aims to leverage the advantages of different base models by inputting their predictions into a meta-model.

- A meta-model is a higher-level model that learns how to combine the predictions of multiple base models to generate the final prediction. It determines the optimal weighting and integration of these individual forecasts, often resulting in improved predictive performance compared to using any single model in isolation.

In [ ]:
# Import Required Libraries
!pip install vecstack

In [ ]:
# Import pandas, sklearn, xgboost, and vecstack libraries
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from vecstack import stacking

In [ ]:
# Import necessary libraries
import pandas as pd

# Define the URL of the dataset
link = 'https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data'

# Define column names for the dataset
names = ['Class', 'Alcohol', 'Malic acid', 'Ash',
         'Alcalinity of ash', 'Magnesium', 'Total phenols',
         'Flavanoids', 'Nonflavanoid phenols', 'Proanthocyanins',
         'Color intensity', 'Hue', 'OD280/OD315 of diluted wines',
         'Proline']

# Read the dataset into a DataFrame
df = pd.read_csv(link, header=None, names=names)

# Display the first 5 rows of the DataFrame
df.head(5)

In [ ]:
# Extract target variable 'Class' into y
y = df[['Class']]

# Extract features into X, excluding 'Class' column
X = df.iloc[:,1:]

# Split dataset into training and testing sets, with 20% for testing
# Use stratified sampling to ensure proportional class distribution
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0, stratify=y)

In [ ]:
# Define a list of classifier models to be used in the ensemble
models = [
    # Model 1: K-Nearest Neighbors Classifier with 5 neighbors
    KNeighborsClassifier(n_neighbors=5, n_jobs=-1),

    # Model 2: Random Forest Classifier with 100 estimators and max depth of 5
    RandomForestClassifier(random_state=0, n_jobs=-1, n_estimators=100, max_depth=5),

    # Model 3: XGBoost Classifier with learning rate 0.1, 100 estimators, and max depth of 5
    XGBClassifier(random_state=0, n_jobs=-1, learning_rate=0.1, n_estimators=100, max_depth=5)
]


__Observation:__
- The different classifiers are set.

__S Train__ refers to a subset of the original training data used for training base models in a stacking ensemble.

__S Test__ is another subset used to generate base model predictions, serving as input for the meta-model. It is crucial for generating predictions from these base models, which then act as new features for the meta-model.

This two-step approach allows the ensemble to capture and leverage the unique strengths of each base model, thus improving overall predictive performance.

- Let us do the S_train and S_tests, as you will be stacking the models by passing train of X and y.

In [ ]:
# Import necessary libraries
from sklearn.preprocessing import LabelEncoder
import warnings

# Suppress warnings related to user warnings and future warnings
warnings.filterwarnings("ignore", category=UserWarning)
warnings.filterwarnings("ignore", category=FutureWarning)

# Initialize LabelEncoder
le = LabelEncoder()

# Transform target variable y_train using LabelEncoder
y_train = le.fit_transform(y_train)

# Perform stacking
S_train, S_test = stacking(models, X_train, y_train, X_test, regression=False, mode='oof_pred_bag',
                           needs_proba=False, save_dir=None, metric=accuracy_score, n_folds=4,
                           stratified=True, shuffle=True, random_state=0, verbose=2)

# Reset warnings to default behavior
warnings.filterwarnings("default", category=UserWarning)
warnings.filterwarnings("default", category=FutureWarning)

__Observations:__
- You can observe the accuracy score and other metrics for different models.
- For the k-nearest neighbor classifier, the mean accuracy is 72%.
- For the random forest, the mean accuracy is 97%.
- For the XGBoost classifier, the mean accuracy is 95%.

### __4.3.1 Advantages of Stacking__

- Accomplishes greater accuracy than simplistic ensemble techniques
- Enables diversity among models
- Possesses the capability to rectify errors made by base models

### __4.3.2 Disadvantages of Stacking__

- Is more challenging to implement and comprehend
- Poses a risk of overfitting the meta-model
- Requires careful selection of both base and meta-models



### __4.4 Blending Ensemble__

Blending is similar to stacking but uses a validation set to train the meta-model. The training set is split into a training and validation set. The base models are trained on the training set, and their predictions for the validation set are used to train the meta-model.

### __4.4.1 Advantages of Blending__

- It reduces the risk of overfitting the meta-model since it is trained on a separate validation set.
- It simplifies the implementation compared to Stacking.

### __4.4.2 Disadvantages of Blending__

- Less data is used to train the base models since the dataset is split.
- The performance can be sensitive to how the data is split.

## __5. AdaBoost Boosting Algorithms__

AdaBoost, an abbreviation for Adaptive Boosting, stands as one of the leading boosting algorithms with widespread adoption. It focuses on classification problems, aiming to transform a collection of weak classifiers into a single strong classifier. It increases the weights of misclassified instances, directing subsequent classifiers to pay more attention to challenging cases.

The AdaBoost algorithm follows these steps:
1. Initially, the algorithm assigns equal weights to all observations in the dataset.
2. A model is constructed using a subset of the data.
3. Predictions are made on the entire dataset using this model.
4. The algorithm computes errors by comparing these predictions to the actual values.
5. In the subsequent model creation, the algorithm assigns higher weights to incorrectly predicted data points.
6. Weights are determined based on error values, where higher errors result in heavier observation weights.
7. This iterative process continues until the error function stabilizes or the algorithm reaches the maximum limit of estimators.

In [ ]:
# Import necessary libraries
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import KFold, cross_val_score
from sklearn.ensemble import BaggingClassifier, AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier

In [ ]:
# Load the dataset from the given URL and assign column names
data = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/breast-cancer-wisconsin.data', header=None)
data.columns = ['Sample code', 'Clump Thickness', 'Uniformity of Cell Size', 'Uniformity of Cell Shape',
                'Marginal Adhesion', 'Single Epithelial Cell Size', 'Bare Nuclei', 'Bland Chromatin',
                'Normal Nucleoli', 'Mitoses','Class']

In [ ]:
# Display first few rows
data.head()

In [ ]:
# Display dataset information
data.info()

In [ ]:
# Drop the 'Sample code' column as it's not required for prediction
data.drop(['Sample code'], axis=1, inplace=True)
data.describe()

In [ ]:
# Check value counts for 'Bare Nuclei' column and replace missing values with 0
# Replacing '?' with 0 handles missing data, ensuring numerical analysis accuracy and avoiding calculation errors

data['Bare Nuclei'].value_counts()
data.replace('?', 0, inplace=True)

In [ ]:
# Convert the 'Bare Nuclei' column to an integer type to ensure a uniform data type across the column,
# which is essential for consistent data manipulation and analysis.
data['Bare Nuclei'] = data['Bare Nuclei'].astype('int64')
data.info()

In [ ]:
# Extract feature matrix X and target vector y
values = data.values

- A SimpleImputer handles missing values in a dataset by imputing (i.e., filling in) the missing values with a specified statistic, such as the mean, median, or most frequent value of the respective feature.
- This is important because missing values can disrupt the analysis and modeling process, leading to biased results or errors in predictive models.
- Imputation helps maintain the integrity and completeness of the dataset, ensuring that the analysis or modeling algorithms can effectively utilize the available data for accurate insights or predictions.

In [ ]:
# Import required libraries
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import MinMaxScaler

# Create instance of SimpleImputer to impute missing values
imputer = SimpleImputer()

# Impute missing values in dataset, obtain imputed data
imputeData = imputer.fit_transform(values)

# Create instance of MinMaxScaler to scale features to range (0, 1)
scaler = MinMaxScaler(feature_range=(0, 1))

# Scale features in imputed data to range (0, 1), obtain normalized data
normalizeddata = scaler.fit_transform(imputeData)

# Split data into features (X) and target variable (y)
X = normalizeddata[:, 0:9]
y = normalizeddata[:, 9]

In [ ]:
# Define K-fold cross-validation
kfold = KFold(n_splits=10, random_state=7, shuffle=True)

In [ ]:
# Bagging classifier with Decision Tree base estimator
cart = DecisionTreeClassifier()
num_trees = 100
model = BaggingClassifier(estimator=cart, n_estimators=num_trees, random_state=7)
results = cross_val_score(model, X, y, cv=kfold)
print(results.mean())

__Observation:__
- With the basic models of BaggingClassifier and DecisionTree, you have an accuracy of 95%.

In [ ]:
# AdaBoost classifier with Decision Tree base estimator
seed = 7
num_trees = 70
model = AdaBoostClassifier(n_estimators=num_trees, random_state=seed)
results = cross_val_score(model, X, y, cv=kfold)
print(results.mean())

__Observations:__
- Both Bagging and AdaBoost use the Decision Tree as a base estimator.
- AdaBoost's higher mean accuracy score compared to Bagging indicates its superior effectiveness in improving the performance of the base estimator.
- This demonstrates the importance and effectiveness of AdaBoost in enhancing the predictive capability of weak learners, like Decision Trees.

## __Conclusion__

This lesson explores various ensemble learning techniques and emphasizes their significance in enhancing the performance of machine learning models. From simple approaches to advanced techniques, you've demonstrated how combining multiple models can lead to improved accuracy, robustness, and generalization.

Specifically, you highlighted the AdaBoost algorithm's ability to transform weak classifiers into strong ones, showcasing its effectiveness in dealing with classification problems. Through examples and practical implementations, this lesson underscores the pivotal role of ensemble learning in achieving more reliable and accurate predictions across diverse applications.